src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+internal sealed class ExtractToComponentBehindCodeActionProvider : IRazorCodeActionProvider
+{
+    private readonly ILogger _logger;
+
+    public ExtractToComponentBehindCodeActionProvider(ILoggerFactory loggerFactory)
+    {
+        if (loggerFactory is null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        _logger = loggerFactory.GetOrCreateLogger<ExtractToComponentBehindCodeActionProvider>();
+    }
+    public Task<IReadOnlyList<RazorVSInternalCodeAction>?> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!context.SupportsFileCreation)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!FileKinds.IsComponent(context.CodeDocument.GetFileKind()))
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, true);
+        if (owner is null)
+        {
+            _logger.LogWarning($"Owner should never be null.");
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var componentNode = owner?.FirstAncestorOrSelf<MarkupElementSyntax>();
+
+        // Make sure we've found tag
+        if (componentNode is null || componentNode.Kind != SyntaxKind.MarkupElement)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        // Do not provide code action if the cursor is not inside a tag
+        if (context.Location.AbsoluteIndex > componentNode.StartTag.Span.End &&
+            context.Location.AbsoluteIndex < componentNode.EndTag.SpanStart)
```

Comment:
This logic seems a bit backwards. For "inside a tag" shouldn't it be checking against the start of the start tag, and the end of the end tag? It's doing the opposite.

Reply:
I might've described the logic incorrectly, although it has the same effect regardless. The code action will not be provided if the cursor is inside html content (or text), that is, after a starting tag and before an ending tag.  This doesn't conflict with other code actions, as the check for whether the componentNode is a markupElement comes just before these lines.

Reply:
I think I understand what you're saying. Would be good to have tests to verify what happens at the very edges of start tags and end tags, but perhaps more importantly, it would be good to expand the comment to be a bit clearer about what is happening, and why the logic seems backwards, even though it isn't. In other words, take most of your reply in this thread, and paste it into the code :)